internal/server/text_synchronization.go

```diff
@@ -0,0 +1,355 @@
+			}
+		}
+	}()
+
+	return nil
+}
+
+// changedText processes document content changes from the client.
+// It supports two modes of operation:
+// 1. Full replacement: Replace the entire document content (when only one change with no range is provided)
+// 2. Incremental updates: Apply specific changes to portions of the document
+//
+// Returns the updated document content or an error if the changes couldn't be applied.
+func (s *Server) changedText(uri string, changes []protocol.TextDocumentContentChangeEvent) ([]byte, error) {
+	if len(changes) == 0 {
+		return nil, fmt.Errorf("%w: no content changes provided", jsonrpc2.ErrInternal)
+	}
+
+	// Check if the client sent the full content of the file.
+	// We accept a full content change even if the server expected incremental changes.
+	if len(changes) == 1 && changes[0].Range == nil && changes[0].RangeLength == 0 {
+		// Full replacement mode
+		return []byte(changes[0].Text), nil
+	}
+
+	// Incremental update mode
+	return s.applyIncrementalChanges(uri, changes)
+}
+
+// applyIncrementalChanges applies a sequence of changes to the document content.
+// For each change, it:
+// 1. Computes the byte offsets for the specified range
+// 2. Verifies the range is valid
+// 3. Replaces the specified range with the new text
+//
+// Returns the updated document content or an error if the changes couldn't be applied.
+func (s *Server) applyIncrementalChanges(path string, changes []protocol.TextDocumentContentChangeEvent) ([]byte, error) {
+	// Get current file content
+	file, ok := s.getProj().File(path)
+	if !ok {
+		return nil, fmt.Errorf("%w: file not found", jsonrpc2.ErrInternal)
+	}
+
+	content := file.Content
+
+	// Apply each change sequentially
+	for _, change := range changes {
+		// Ensure the change includes range information
+		if change.Range == nil {
+			return nil, fmt.Errorf("%w: unexpected nil range for change", jsonrpc2.ErrInternal)
+		}
+
+		// Convert LSP positions to byte offsets
+		start := s.PositionOffset(content, change.Range.Start)
+		end := s.PositionOffset(content, change.Range.End)
+
+		// Validate range
+		if end < start {
+			return nil, fmt.Errorf("%w: invalid range for content change", jsonrpc2.ErrInternal)
+		}
+
+		// Apply the change
+		var buf bytes.Buffer
+		buf.Write(content[:start])
+		buf.WriteString(change.Text)
+		buf.Write(content[end:])
+		content = buf.Bytes()
+	}
+
+	return content, nil
+}
+
+// PositionOffset converts an LSP position (line, character) to a byte offset in the document.
+// It calculates the offset by:
+// 1. Finding the starting byte offset of the requested line
+// 2. Adding the character offset within that line, converting from UTF-16 to UTF-8 if needed
+//
+// Parameters:
+// - content: The file content as a byte array
+// - position: The LSP position with line and character numbers (0-based)
+//
+// Returns the byte offset from the beginning of the document
+func (s *Server) PositionOffset(content []byte, position Position) int {
```

Comment:
This seems more like a func rather than a method of `Server`, considering it doesn't actually need to access `Server`. And, if it's made a func, I think placing it in `internal/server/util.go` would be more appropriate.